src/GitHub.Services.Vssdk/AsyncMenuPackage.cs

```diff
@@ -0,0 +1,41 @@
+﻿using System;
+using System.Threading;
+using System.ComponentModel.Design;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Task = System.Threading.Tasks.Task;
+
+namespace GitHub.Services.Vssdk
+{
+    public abstract class AsyncMenuPackage : AsyncPackage
+    {
+        IVsUIShell vsUIShell;
+
+        sealed protected async override Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
+        {
+            vsUIShell = await GetServiceAsync(typeof(SVsUIShell)) as IVsUIShell;
```

Comment:
You should use `await this.JoinableTaskFactory.SwitchToMainThreadAsync(CancellationToken);` as the first line in your method. Casting the SVsUIShell object to an interface itself generally requires the UI thread, and otherwise you're vulnerable to a deadlock by relying on RPC to get you to the UI thread.

Reply:
There is an helper method btw that you can use as ServiceProvider.GetGlobalServiceAsync<SVsUIShell, IVsUIShell> that handles this automatically.